### PR TITLE
add a note to `@var = T.cast(...)` errors to suggest a way forward

### LINKED ISSUE #4017: Confusing error using `T.cast` on ivar
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20A%0A%20%20extend%20T%3A%3ASig%0A%20%20def%20initialize%0A%20%20%20%20%40val%20%3D%20T.cast(0%2C%20Integer)%0A%20%20end%0Aend)

```ruby
# typed: true

class A
  extend T::Sig
  def initialize
    @val = T.cast(0, Integer)
  end
end
```

#### Observed output

```
editor.rb:6: Use T.let to specify the type of constants https://srb.help/5013
     6 |    @val = T.cast(0, Integer)
                   ^^^^^^^^^^^^^^^^^^
Errors: 1
```

#### Expected behavior

https://srb.help/5013 should mention the workaround for when `T.cast` is intended:

```ruby
val = T.cast(x, X)
@val = val
```

we might also want to use `addErrorNote` on the error message to suggest something like "if you really meant to use `T.cast`, do this...".

### PR DESCRIPTION
**Pull request by froydnj**
### Motivation

Fixes #4017.

We updated the docs for this (https://srb.help/5013), presumably after the original bug was filed.  But it's nice to save somebody a click.  (I don't know how to do an autocorrect that involves temporary variables, so I opted to leave that up to the user -- and they may want to use `T.let` in any event.)

### Test plan

See included automated tests.

### CODE REVIEW

File: resolver/resolver.cc [hunk 1]
```diff
@@ -1731,6 +1731,12 @@ class ResolveTypeMembersAndFieldsWalk {
         } else if (cast->cast != core::Names::let()) {
             if (auto e = ctx.beginError(cast->loc, core::errors::Resolver::ConstantAssertType)) {
                 e.setHeader("Use `{}` to specify the type of constants", "T.let");
+                if (cast->cast == core::Names::cast()) {
+                    e.addErrorNote("If you really want to use `T.cast`, "
+                                   "assign to an intermediate variable first and then "
+                                   "assign that variable to `{}`",
+                                   asgn.lhs.toString(ctx));
+                }
```
Comment (elliottt) [inline resolver/resolver.cc:1739]: This seems reasonable to me, but what do you think about adding an autocorrect as well?

**Review by elliottt — APPROVED** [review]
Thanks for adding this! Only one non-blocking comment about perhaps making it an autocorrect as well.

File: resolver/resolver.cc (hunk 1 above)
Comment (froydnj) [reply resolver/resolver.cc:1739]: I'm happy to add one.

Do we have a policy on adding autocorrects for things?  My assumption was that we only have autocorrects for things that are "obviously" the right thing to do, but adding an autocorrect here is more like "probably" the right thing to do.

**Comment by froydnj** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JvLZ4Wj6h8w77c
→ https://go/builds/bui_JvLZwLBQWQTRhp

**Pull request merged**